File: 15-3sum/15-3sum.scala
object Solution {
  def threeSum(nums: Array[Int]): List[List[Int]] = {
    var result = List[List[Int]]()
    val len = nums.length
    if (len < 3) {
      return result
    }
    val sorted = nums.sorted
    if (sorted(0) > 0 || sorted(len-1) < 0) {
      return result
    }    
    if (sorted(0) == 0 && sorted(len-1) == 0) {
      result ::= List(0, 0, 0)
      return result
    }
    var i = 0
    while (i < len-2) {
      val a = sorted(i)
      if (a > 0) {
        return result
      }
      var l = i + 1
      var r = len - 1
      while (l < r) {
        val b = sorted(l)
        val c = sorted(r)
        if (c < 0) {
          return result
        }
        if (a+b+c < 0) {
          do {
            l += 1
          } while (l < r && b == sorted(l))
        } else if (a+b+c == 0) {
          result ::= List(a, b, c)
          do {
            l += 1
          } while (l < r && b == sorted(l))
          do {
            r -= 1
          } while (l < r && c == sorted(r))
        } else {
          do {
            r -= 1
          } while (l < r && c == sorted(r))
        }
      }
      do {
        i += 1
      } while (i < len-2 && a == sorted(i))
    }
    return result
  }
}

File: 15-3sum/15-3sum.go
func threeSum(nums []int) [][]int {
  sort.Ints(nums)
  length := len(nums)
  if length < 3 { return [][]int{} }
  ans := make([][]int, 0, length - 2)
  ni := nums[0] - 1
  for i := 0; i < length - 2; i++ {
    for i < length - 2 && nums[i] == ni { i++ }
    ni = nums[i]
    if ni > 0 { break }
    nj := nums[i+1] - 1
    for j := i + 1; j < length - 1; j++ {
      for j < length - 1 && nums[j] == nj { j++ }
      nj = nums[j]
      target := 0 - ni - nj
      if nj > target { break }
      l, r := j+1, length - 1
      for l <= r {
        m := (l+r) / 2
        nk := nums[m]
        if nk == target {
          
          ans = append(ans, []int{ni, nj, nk})
          break
        } else if nk < target {
          l = m + 1
        } else {
          r = m - 1
        }
      }
    }
  }
  return ans
}